### PR TITLE
Rework M.2 support

### PR DESCRIPTION
**Pull request by jan-kiszka**
This changes the interface between firmware and kernel for configuring
and communicating the M.2 mode.

Previously, a different device tree was selected by the firmware when a
different mode was set, including repeating updates of the U-Boot
environment in flash on every boot. This approach here uses the device
tree fix-up of U-Boot to adjust a given DT to the selected mode.

Further changes:
 - permit manual configuration in secure boot mode
 - report selected mode and if it was chosen manually or as fallback
 - align U-Boot DT to kernel DT
 - document manual configuration
 - cleanups and simplifications

### CODE REVIEW

File: recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch [hunk 1]
```diff
@@ -114,14 +122,34 @@ index 0000000000..741b9177da
 +	>;
 +};
 +
++/*
++ * Default configuration for B-key with 2 PCIe lanes, E-key disabled.
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch:126]: Default is B-key with usb3.0 and E-key with pcie one lane.
even if B-key with two lane, E-key could also use 2.0

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
++				 <&k3_clks 151 8>;  /* set PIPE3_TXB_CLK to WIZ8B2M4VSB */
++	phys = <&serdes0 PHY_TYPE_USB3 0>;
++	phy-names = "usb3-phy";
++};
++
++&usb0 {
++	maximum-speed = "super-speed";
++	snps,dis-u1-entry-quirk;
++	snps,dis-u2-entry-quirk;
++};
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:193]: As mentioned above, default should be bkey usb3.0 + ekey pcie

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
++static void m2_overlay_prepare(void)
++{
++#if defined(CONFIG_OF_LIBFDT) && defined(CONFIG_OF_BOARD_SETUP)
++	const char *overlay_path;
++	void *overlay;
++	u64 loadaddr;
++	ofnode node;
++	int ret;
++
++	if (connector_mode == BKEY_PCIEX2_EKEY_NONE)
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:356]: same as above

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
++					       CONNECTOR_MODE_INVALID);
++	const char *mode_info = "";
++	struct m2_config_pins config_pins;
++	unsigned int n;
++
++	/* enable M.2 connector power */
++	set_pinvalue("gpio@601000_17", "P3V3_M2_EN", 1);
++	udelay(4 * 100);
++
++	m2_manual_config = env_get_ulong("m2_manual_config", 10,
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:402]: 392 line, twice read

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+From: Jan Kiszka <jan.kiszka@siemens.com>
+Date: Tue, 11 Oct 2022 13:23:21 +0200
+Subject: [PATCH 28/32] iot2050: Add support for configuring M.2 connector
+
+The M.2 slots of the related IOT2050 variant need to be configured
+according to the plugged cards. This tries to detect the card using the
+M.2 configuration pins of the B-key slot. If that fails, a U-Boot
+environment variable can be set to configure manually. This variable is
+write-permitted also in secure boot mode as it is not able to undermine
+the integrity of the booted system.
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:11]: manual configuration prior to auto-detection.
so it's not "If that fails, a U-Boot environment variable can be set to configure manually"

whether or not fails, manual configuration should be first priority if existed

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
++
++	/* enable M.2 connector power */
++	set_pinvalue("gpio@601000_17", "P3V3_M2_EN", 1);
++	udelay(4 * 100);
++
++	m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					 CONNECTOR_MODE_INVALID);
++	if (m2_manual_config < CONNECTOR_MODE_INVALID) {
++		mode_info = " [manual mode]";
++		connector_mode = m2_manual_config;
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:406]: should check range

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
++			if (!memcmp(config_pins.config,
++				    m2_config_table[n].config_pins.config,
++				    sizeof(config_pins.config))) {
++				connector_mode = m2_config_table[n].mode;
++				break;
++			}
++		}
++		if (connector_mode == CONNECTOR_MODE_INVALID) {
++			mode_info = " [fallback, card unknown/unsupported]";
++			connector_mode = BKEY_USB30_EKEY_PCIE;
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:423]: this is the real default mode.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
++		for (n = 0; n < ARRAY_SIZE(config_pins.config); n++)
++			config_pins.config[n] =
++				get_pinvalue(m2_bkey_cfg_pin_info[n].gpio_name,
++					     m2_bkey_cfg_pin_info[n].label);
++		connector_mode = CONNECTOR_MODE_INVALID;
++		for (n = 0; n < ARRAY_SIZE(m2_config_table); n++) {
++			if (!memcmp(config_pins.config,
++				    m2_config_table[n].config_pins.config,
++				    sizeof(config_pins.config))) {
++				connector_mode = m2_config_table[n].mode;
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:417]: mode_info should also be set

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
++	int err;
++
++	if (!connector_overlay)
++		return;
++
++	overlay = malloc(connector_overlay_size);
++	if (!overlay)
++		goto fixup_error;
++
++	memcpy(overlay, connector_overlay, connector_overlay_size);
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:473]: why memcpy again？

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
++	err = fdt_overlay_apply_verbose(blob, overlay);
++	free(overlay);
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:475]: two malloc one free

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
++
++	if (!err)
++		return;
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:478]: no error print?

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
++----------------------
++
++The M.2 variant of the IOT2050 comes with one B-key and one E-key slot. These
++are configured by U-Boot depending on the detected usage (auto configuration).
++The device tree loaded later on for the OS will then be fixed up according to
++this configuration.
++
++If this auto configuration does not work reliably, the U-Boot environment
++variable "m2_manual_config" can be set to select the mode manually:
++
```
Comment (AsuraZeng) [inline recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:516]: better to add: manual config always prior to auto-detect

File: recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
++&main_gpio0 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&main_m2_pcie_mux_control>;
++};
++
++&main_gpio1 {
++	pinctrl-names = "default";
++	pinctrl-0 = <
++	    &main_m2_enable_pins_default
++		&main_pmx0_m2_config_pins_default
```
Comment (AsuraZeng) [inline recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch:117]: indention

File: recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
++		&main_pmx1_m2_config_pins_default
++	>;
++};
++
++/*
++ * Default configuration for B-key with 2 PCIe lanes, E-key disabled.
```
Comment (AsuraZeng) [inline recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch:123]: default is bkey usb3.0 ekey pcie

**Comment by AsuraZeng** [comment]
why no dtbo in kernel?

**Comment by AsuraZeng** [comment]
in secure mode, ebg will load the overlay or not?

File: recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch (hunk 1 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch:126]: Nope, this was changed in order to make overlays possible: U-Boot does not support removing properties via overlays. Therefore, the baseline DT must start in a way that we only add or change properties.

Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch:126]: Oh, that the E-key can have a role when the B-key uses two PCI lanes, that was nowhere mentioned so far. Can you elaborate please?

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 2 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:193]: Nope.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 5 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:11]: It is meant as "if auto-configuration does not work for you, you can set up manual configuration". I will try to reword this to make it clearer.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 4 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:402]: oops

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 5 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:11]: "For the case auto configuration does not work reliably, it is possible to set
the U-Boot environment variable "m2_manual_config" to select the mode manually:"

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 6 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:406]: Connector mode can be any value if the user does not read the documentation.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 7 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:423]: It's changed, as I wrote.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 9 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:473]: Apply the overlay will unfortunately modify that blob. So we need to preserve the original for future fix-ups.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 10 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:475]: The first malloc is (almost) persistent: we only free the first copy implicitly when U-Boot boot-time services terminate (on hand-over to OS during legacy boot or when UEFI boot-time services are terminated). Thus no free.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 11 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:478]: We fall through to it.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 12 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:516]: See above.

File: recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch (hunk 13 above)
Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch:117]: Fixed

File: recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch (hunk 14 above)
Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-5.10/0120-arm64-dts-ti-iot2050-Add-support-for-M.2-variant.patch:123]: Nope.

**Comment by jan-kiszka** [comment]
> why no dtbo in kernel?

The kernel has not use case for that overlay.

**Comment by jan-kiszka** [comment]
> in secure mode, ebg will load the overlay or not?

No, firmware will load it (it's embedded) and apply it when the UEFI bootloader (EBG, systemd-boot) requests EFI_DT_FIXUP_PROTOCOL.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 8 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:417]: Empty mode_info implies "auto" so far. I don't think we need to write this out explicitly.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 6 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:406]: Oh, misread as "why are we checking the range" - we are checking it to be >= 0 and < CONNECTOR_MODE_INVALID.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 11 above)
Comment (AsuraZeng) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:478]: misread the err.

**Comment by AsuraZeng** [comment]
> > why no dtbo in kernel?
> 
> The kernel has not use case for that overlay.

yes, but we intend to overlay kernel's dtb.However the dtb is produced under uboot, I think it's better to place copies in the kernel also.

**Comment by jan-kiszka** [comment]
>

**Comment by jan-kiszka** [comment]
> > > why no dtbo in kernel?
> > 
> > 
> > The kernel has not use case for that overlay.
> 
> yes, but we intend to overlay kernel's dtb.However the dtb is produced under uboot, I think it's better to place copies in the kernel also.

Again: what for?

File: recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch (hunk 1 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch:126]: OK, will update the comment and outputs accordingly.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 8 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:417]: We write out the detected mode already.

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 10 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:475]: Then we hard-code in the code the max size of the overlay.

File: recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch (hunk 1 above)
Comment (BaochengSu) [reply recipes-bsp/u-boot/files/0027-arm-dts-iot2050-Add-support-for-M.2-variant.patch:126]: The title in iot2050setup should be refined to `M.2 PCIE/USB3.0 Configuration`

**Comment by AsuraZeng** [comment]
> > > > why no dtbo in kernel?
> > > 
> > > 
> > > The kernel has not use case for that overlay.
> > 
> > 
> > yes, but we intend to overlay kernel's dtb.However the dtb is produced under uboot, I think it's better to place copies in the kernel also.
> 
> Again: what for?

Can this dtbo be used in the kernel? Put dtbo in the file system.
The advantage of this is that there is no need to forcibly bind the dtbo of the kernel to uboot.
The disadvantage may also require additional consideration of the secure aspect

File: recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch (hunk 9 above)
Comment (BaochengSu) [reply recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch:473]: Find for this round. However, better to add comments to this memory area so that people who do not know the details would not treat this area as redundant and delete it accidentally, this could be done in the upstreaming patch.

**Review by BaochengSu — APPROVED** [review]

**Comment by jan-kiszka** [comment]
I'll also prepare an MR for master then, or should I wait for further testing? Ordering is unusual already anyway.

**Comment by chombourger** [comment]
There is unfortunately no PV or PR bump in this changeset meaning that the linux-image-iot2050 package will have the same version and will not be upgraded in the `apt dist-upgrade` scenario

**Comment by jan-kiszka** [comment]
Right, we do not address package updates so far, specifically of the kernel as it would not work under all circumstances (secure boot). That said, we can discuss where such a versioning policy would help, how is should look like and how we make sure to enforce it.
